[CSS] Allow nesting style rule with relaxed syntax

mdubet opened the pull request:
#### 2cfe184f6dbee766aefa51db1de3e613c185b0fb
<pre>
[CSS] Allow nesting style rule with relaxed syntax
<a href="https://bugs.webkit.org/show_bug.cgi?id=260812">https://bugs.webkit.org/show_bug.cgi?id=260812</a>
rdar://113475843

Reviewed by Antti Koivisto.

<a href="https://drafts.csswg.org/css-nesting/#nesting">https://drafts.csswg.org/css-nesting/#nesting</a>

The CSS nesting spec now allows having identifier as the first token
in the nested style rule selector.
This implementation follows the spec algorithm of first trying to parse
a declaration, then a nested style rule, otherwise it goes to error recovery (until next semicolon).

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/parsing-expected.txt:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeStyleRule):
(WebCore::CSSParserImpl::consumeBlockContent):
(WebCore::CSSParserImpl::consumeDeclarationListInNewNestingContext):
(WebCore::CSSParserImpl::consumeDeclarationList):
(WebCore::CSSParserImpl::consumeStyleBlock):
(WebCore::CSSParserImpl::consumeDeclaration):
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper): Deleted.
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumeNestedSelectorList):

Canonical link: <a href="https://commits.webkit.org/267549@main">https://commits.webkit.org/267549@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aa294dedf3abc1b188ddcaf52560f21e380dfe5a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16928 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17251 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17710 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18711 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15847 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17118 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20483 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17389 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17127 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17485 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14657 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19521 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14728 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15352 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22073 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15712 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15519 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19834 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16125 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13662 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15288 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4056 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19652 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15955 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -1272,7 +1284,10 @@ void CSSParserImpl::consumeDeclaration(CSSParserTokenRange range, StyleRuleType
+    const size_t propertiesCount = topContext().m_parsedProperties.size();
+    auto isParsed = [&] {
+        return topContext().m_parsedProperties.size() != propertiesCount;
+    };
```
I think `propertiesCount` is like `oldPropertyCount` and `isParsed()` is `didParseNewProperties()` or `isValidDeclaration()` or similar.

Shouldn't `propertiesCount` be always zero here in any case?

mdubet replied:
Indeed, I've renamed those.

`propertiesCount` can be non zero though (this parses a single property, it will be non zero on the second property in the same scope/rule).

anttijk commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -1150,29 +1150,52 @@ void CSSParserImpl::consumeDeclarationListOrStyleBlockHelper(CSSParserTokenRange
 
-            consumeDeclaration(range.makeSubRange(declarationStart, &range.peek()), ruleType);
+            const auto declarationRange = range.makeSubRange(declarationStart, &range.peek());
+            auto valid = consumeDeclaration(declarationRange, ruleType);
```
boolean variable are usually named with is/was/has or similar so this could be `isValidDeclaration`.

anttijk commented on LayoutTests/fast/table/frame-and-rules.html:
```diff
@@ -1 +1 @@
	</tfoot>
</table>
</body>
</html>
```
what's up with this test?

mdubet replied:
It was part of the test typos fixing.
I've removed all of them from this patch.

mdubet committed:
[CSS] Allow nesting style rule with relaxed syntax
https://bugs.webkit.org/show_bug.cgi?id=260812
rdar://113475843

Reviewed by Antti Koivisto.

https://drafts.csswg.org/css-nesting/#nesting

The CSS nesting spec now allows having identifier as the first token
in the nested style rule selector.
This implementation follows the spec algorithm of first trying to parse
a declaration, then a nested style rule, otherwise it goes to error recovery (until next semicolon).

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/parsing-expected.txt:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeStyleRule):
(WebCore::CSSParserImpl::consumeBlockContent):
(WebCore::CSSParserImpl::consumeDeclarationListInNewNestingContext):
(WebCore::CSSParserImpl::consumeDeclarationList):
(WebCore::CSSParserImpl::consumeStyleBlock):
(WebCore::CSSParserImpl::consumeDeclaration):
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper): Deleted.
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumeNestedSelectorList):

Canonical link: https://commits.webkit.org/267549@main
